### PR TITLE
[LV] Linear argument tests for vectorization of function calls

### PR DESCRIPTION
**Pull request by huntergr-arm**
Tests to exercise vectorization of function calls where a vector variant takes a linear parameter.

### CODE REVIEW

File: llvm/test/Transforms/LoopVectorize/AArch64/vector-call-linear-args.ll [hunk 1]
```diff
@@ -0,0 +1,275 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 2
+; RUN: opt < %s -passes=loop-vectorize,instsimplify -force-vector-interleave=1 -S | FileCheck %s
```
Comment (mgabka) [inline llvm/test/Transforms/LoopVectorize/AArch64/vector-call-linear-args.ll:2]: Hi @huntergr-arm ,
wouldn't be better to remove the "test_linear_with_mask" test function, but:
* have more than 1 mapping available per scalar call, (as that is most likely going to be real life scenario), so we can test if LV is choosing the right one,
* add more RUN lines?
1. RUN: opt < %s -passes=loop-vectorize,instsimplify -force-vector-interleave=1 //only Neon mappings used
2.RUN: opt < %s -passes=loop-vectorize,instsimplify -force-vector-interleave=1 -mattr=+sve //only sve mappings used
3.RUN: opt < %s -passes=loop-vectorize,instsimplify -force-vector-interleave=1 -mattr=+sve -prefer-predicate-over-epilogue=predicate-dont-vectorize //sve mappings used, but also Neon masked mapping used when available

Comment (mgabka) [reply llvm/test/Transforms/LoopVectorize/AArch64/vector-call-linear-args.ll:2]: I also think we should add at least one test case where step is negative, as that is allowed I think. 
The VFABI mangling allows it, from the specification: "For example, linear(val(i):-3) gives Ln3 when the type of i is integer"

**Comment by huntergr-arm** [comment]
I added a filter to the check line updating to focus on the call instructions generated for each RUN line.

File: llvm/test/Transforms/LoopVectorize/AArch64/vector-call-linear-args.ll [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+  store i64 %data, ptr %gepa
+  %indvars.iv.next = add nuw nsw i64 %indvars.iv, 1
+  %exitcond = icmp eq i64 %indvars.iv.next, %n
+  br i1 %exitcond, label %for.cond.cleanup, label %for.body
+
+for.cond.cleanup:
+  ret void
+}
+
+define void @test_linear_with_vector(ptr noalias %a, ptr readnone %b, ptr readonly %c, i64 %n) {
```
Comment (mgabka) [inline llvm/test/Transforms/LoopVectorize/AArch64/vector-call-linear-args.ll:40]: **test_vector_linear4**

File: llvm/test/Transforms/LoopVectorize/AArch64/vector-call-linear-args.ll [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+  store i64 %data, ptr %gepa
+  %indvars.iv.next = add nuw nsw i64 %indvars.iv, 1
+  %exitcond = icmp eq i64 %indvars.iv.next, %n
+  br i1 %exitcond, label %for.cond.cleanup, label %for.body
+
+for.cond.cleanup:
+  ret void
+}
+
+define void @test_linear_wide_stride(ptr noalias %a, ptr readnone %b, i64 %n) {
```
Comment (mgabka) [inline llvm/test/Transforms/LoopVectorize/AArch64/vector-call-linear-args.ll:102]: test_linear16_wide_stride

File: llvm/test/Transforms/LoopVectorize/AArch64/vector-call-linear-args.ll [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+  store i64 %data, ptr %gepa
+  %indvars.iv.next = add nuw nsw i64 %indvars.iv, 1
+  %exitcond = icmp eq i64 %indvars.iv.next, %n
+  br i1 %exitcond, label %for.cond.cleanup, label %for.body
+
+for.cond.cleanup:
+  ret void
+}
+
+define void @test_linear_mixed_types(ptr noalias %a, ptr readnone %b, ptr readonly %c, i64 %n) {
```
Comment (mgabka) [inline llvm/test/Transforms/LoopVectorize/AArch64/vector-call-linear-args.ll:133]: test_linear4_linear8

File: llvm/test/Transforms/LoopVectorize/AArch64/vector-call-linear-args.ll [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+  store i32 %data, ptr %gepa, align 8
+  %indvars.iv.next = add nuw nsw i64 %indvars.iv, 1
+  %exitcond = icmp eq i64 %indvars.iv.next, %n
+  br i1 %exitcond, label %for.cond.cleanup, label %for.body
+
+for.cond.cleanup:
+  ret void
+}
+
+define void @test_linear_non_ptr(ptr noalias %a, i64 %n) {
```
Comment (mgabka) [inline llvm/test/Transforms/LoopVectorize/AArch64/vector-call-linear-args.ll:164]: test_linear3_non_ptr

File: llvm/test/Transforms/LoopVectorize/AArch64/vector-call-linear-args.ll [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+  store i32 %data, ptr %gepa
+  %indvars.iv.next = add nuw nsw i64 %indvars.iv, 1
+  %exitcond = icmp eq i64 %indvars.iv.next, %n
+  br i1 %exitcond, label %for.cond.cleanup, label %for.body
+
+for.cond.cleanup:
+  ret void
+}
+
+define void @test_linear_non_ptr_neg_stride(ptr noalias %a, i64 %n) {
```
Comment (mgabka) [inline llvm/test/Transforms/LoopVectorize/AArch64/vector-call-linear-args.ll:195]: test_linear5_non_ptr_negative_stride

File: llvm/test/Transforms/LoopVectorize/AArch64/vector-call-linear-args.ll [hunk 7]
```diff
@@ -0,0 +1,253 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --filter "call" --version 2
+; RUN: opt < %s -passes=loop-vectorize -force-vector-interleave=1 -S | FileCheck %s --check-prefixes=NEON
+; RUN: opt < %s -mattr=+sve -passes=loop-vectorize -force-vector-interleave=1 -S | FileCheck %s --check-prefixes=SVE_OR_NEON
+; RUN: opt < %s -mattr=+sve -passes=loop-vectorize -force-vector-interleave=1 -S -prefer-predicate-over-epilogue=predicate-dont-vectorize | FileCheck %s --check-prefixes=SVE_TF
```
[the remark this answers is not in the document]
Comment (huntergr-arm) [reply llvm/test/Transforms/LoopVectorize/AArch64/vector-call-linear-args.ll:4]: Only in the predicate-dont-vectorize run.

**Review by mgabka — APPROVED** [review]

**Pull request merged**